### ISSUE TITLE
Rate limit of mount events makes starting/stopping mount units unreliable

### ISSUE THREAD

**Issue by Vogtinator** [systemd/systemd#20329]
**systemd version the issue has been seen with**
248.5 and 248.6, not 248.3

**Used distribution**
 > openSUSE Tumbleweed

**Linux kernel version used** (`uname -a`)

 > `Linux d142 5.13.4-1-default #1 SMP Thu Jul 22 15:55:06 UTC 2021 (91a0cca) x86_64 x86_64 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Unexpected behaviour you saw**
 > After a batch of (un)mount events which trigger the sd-event rate limit, starting/stopping of (unrelated) `.mount` units becomes unreliable.
> Doing a `umount` followed by a `systemctl start` of the `.mount` unit is successful, but the mount didn't actually happen.
> Doing a `mount` followed by a `systemctl stop` of the `.mount` unit is successful, but the mount is still there.

**Steps to reproduce the problem**
This is a somewhat artificial test. In a real system, this is triggered by setup and teardown of a `chroot` environment, which breaks other services running shortly after it.

```
echo 'tmpfs /mnt tmpfs defaults,nofail 0 0' >> /etc/fstab
systemctl daemon-reload
mkdir -p foo
systemctl start mnt.mount

for ((i=0;i<50;++i)); do mount --bind foo foo; umount foo; done; # Trigger rate limit
umount /mnt; systemctl start mnt.mount; echo $?; mountpoint /mnt; systemctl status mnt.mount; mountpoint /mnt
```

It's expected that after `umount /mnt; systemctl start mnt.mount`, the `mnt.mount` unit is active and `/mnt` mounted. What actually happens is that the `systemctl start mnt.mount` effectively does nothing: The `mnt.mount` unit is (still) up, but `/mnt` is not mounted:

```
0
/mnt is not a mountpoint
● mnt.mount - /mnt
     Loaded: loaded (/etc/fstab; generated)
     Active: active (mounted) since Wed 2021-07-28 09:36:10 UTC; 4s ago
      Where: /mnt
       What: tmpfs
       Docs: man:fstab(5)
             man:systemd-fstab-generator(8)
      Tasks: 0 (limit: 4660)
        CPU: 5ms
     CGroup: /system.slice/mnt.mount
/mnt is not a mountpoint
```

There's a clear "happens-before" relationship of `umount` and `systemctl start`, so I would not expect any race condition.

**Additional program output to the terminal or log subsystem illustrating the issue**

Because of the rate limit, the unmount event of `/mnt` is processed after the request to `start mnt.mount`, so at that point systemd believes that the mount is still active and it doesn't actually do the mount. It looks like this in the debug log:

```text
Event source 0x561313e8ad70 (mount-monitor-dispatch) entered rate limit state.
...
mnt.mount: Trying to enqueue job mnt.mount/start/replace
mnt.mount: Installed new job mnt.mount/start as 985
mnt.mount: Enqueued job mnt.mount/start as 985
...
mnt.mount: Job 985 mnt.mount/start finished, result=done
...
Event source 0x561313e8ad70 (mount-monitor-dispatch) left rate limit state.
mnt.mount: Deactivated successfully.
mnt.mount: Changed mounted -> dead
mnt.mount: Consumed 4ms CPU time.
```

What could be done is to force processing of all mount events in the queue whenever a `.mount` unit is queried or changed. That approach should not conflict with #15464.

A workaround is to never mix `systemctl start/stop .mount` and `mount/umount`.

**Comment by poettering** [comment]
What about 249? This piece of code received fixes since 248.

**Comment by Vogtinator** [comment]
Happens as well with v249(.0):

```
d142:~ # for ((i=0;i<50;++i)); do mount --bind foo foo; umount foo; done; umount /mnt; systemctl start mnt.mount; echo $?; mountpoint /mnt; systemctl status mnt.mount; mountpoint /mnt
0
/mnt is not a mountpoint
● mnt.mount - /mnt
     Loaded: loaded (/etc/fstab; generated)
     Active: active (mounted) since Thu 2021-07-29 09:38:01 UTC; 33s ago
      Where: /mnt
       What: tmpfs
       Docs: man:fstab(5)
             man:systemd-fstab-generator(8)
      Tasks: 0 (limit: 4660)
        CPU: 2ms
     CGroup: /system.slice/mnt.mount
/mnt is not a mountpoint
d142:~ # rpm -q systemd
systemd-249-6.1.x86_64
```

**Comment by bluca** [comment]
Isn't this the usual problem with /proc/mountinfo being racy and generally a terrible interface for anything besides the simplest use cases? What happened to the mount notifications and fsinfo() patch series? /cc @brauner

**Comment by brauner** [comment]
> Isn't this the usual problem with /proc/mountinfo being racy and generally a terrible interface for anything besides the simplest use cases? What happened to the mount notifications and fsinfo() patch series? /cc @brauner

It died because of Red Hat internal disagreements what this interface should really look like ending in a shouting match and the patch got abandoned. Though David still intends to work on it. I'll chat to him about it soon.

**Comment by raven-au** [comment]
> > Isn't this the usual problem with /proc/mountinfo being racy and generally a terrible interface for anything besides the simplest use cases? What happened to the mount notifications and fsinfo() patch series? /cc @brauner
> 
> It died because of Red Hat internal disagreements what this interface should really look like ending in a shouting match and the patch got abandoned. Though David still intends to work on it. I'll chat to him about it soon.

This is completely wrong, there was a lot of development and many changes along the way but there weren't internal disagreements preventing it from being merged.

The problem ended up being the fsinfo() system call, the notifications series would probably have been merged (and I think was partly merged but, for the mount table use case that's not enough, the fsinfo() call is also needed.

Linus wanted people that would use the fsinfo() system call and had code ready to go to speak up (kind of unrealistic since those people are probably not reading the LKML or fsdevel mailing lists, but that's the way it was).

Karl Zak and I had been working on code for libmount and I was well along with code for systemd but that wasn't enough for Linus.

The fsinfo() system call had a much larger scope because was meant to provide a wide range of information about file systems and allow file systems to customize the information if the filesystem needed to do so.

It wasn't just about providing mount table information in an efficient way.

In the end we couldn't get enough buy in from a sufficient number of potential users to speak up so Linus wouldn't merge the fsinfo() changes.

AFAIK David did need those changes which is what the problem is now I believe.

Unless we can get sufficient buy in from others that would use this change, for other than the mount table case, then we are stuck.

**Comment by bluca** [comment]
Thanks for the update, very interesting. Any reason why the mount table use case is not enough by itself? I mean we'd use it here without any doubt, this is one of the most long standing and unsolvable issues we have - using mounts at scale and reliably is factually impossible without those kernel changes. Isn't this enough for merging? What other users or use cases should speak up?

**Comment by raven-au** [comment]
> Thanks for the update, very interesting. Any reason why the mount table use case is not enough by itself? I mean we'd use it here without any doubt, this is one of the most long standing and unsolvable issues we have - using mounts at scale and reliably is factually impossible without those kernel changes. Isn't this enough for merging? What other users or use cases should speak up?

I guess that's part of the problem, I don't know what David had in mind or who was asking for it so I couldn't ask people to join in myself. I can see the utility of what was being provided though.

There was a lot of duplication of information provided by existing system calls and a lot of extended information too.
And there's ongoing support, at one time Al said he didn't like the way it was extendible saying it would end up a maintenance headache.

Anyway it ended up Linus essentially saying who's going to use "all" this and not getting the answers he wanted, I believe he wasn't convinced the existing proc mount table handling was as serious a problem as it in spite of the performance improvements for mount handling (I'm pretty sure I describe my findings, it was a while ago). But he did say he got the need for the notifications improvements.

Unfortunately without a relatively low overhead way to get the extra information about a given mount directly the notifications, in particular the event queuing functionality, it isn't really that useful AFAICS.

For my part it is the mount table handling that I need it for too, not so much for autofs itself but for the disruption that can be caused by large mount tables and frequent mount activity due to people using largish autofs configurations. It can play havoc on applications that monitor the mount table, like systemd and udiskd. But it sounds like your no stranger to that.

I'm tempted to bore you with some of the test results but this is already way out of scope for this issue, ;)

**Comment by raven-au** [comment]
Getting back to the problem here though, I always thought that the 1 second quiescence of the rate limiting was way too long.

I wonder if cutting that back to around 250ms would mostly resolve this problem, a sort of a poor mans event queuing facility rather than a let's just ignore these things for a "whole" second to lower overhead. I wonder how much impact that would have on the rate limiting performance improvement too ...

**Comment by raven-au** [comment]
> I guess that's part of the problem, I don't know what David had in mind or who was asking for it so I couldn't ask people to join in myself. I can see the utility of what was being provided though.

I do remember David saying there were AFS use cases he needed it for ... can't remember details though.

**Comment by brauner** [comment]
> > > Isn't this the usual problem with /proc/mountinfo being racy and generally a terrible interface for anything besides the simplest use cases? What happened to the mount notifications and fsinfo() patch series? /cc @brauner
> > 
> > 
> > It died because of Red Hat internal disagreements what this interface should really look like ending in a shouting match and the patch got abandoned. Though David still intends to work on it. I'll chat to him about it soon.
> 
> This is completely wrong, there was a lot of development and many changes along the way but there weren't internal disagreements preventing it from being merged, at least not within RedHat as far as I know.
> 
> You must be referring to Miklos Szeredi not liking the fsinfo() interface which was unfortunate and IMHO missed the point of what the fsinfo() system call was meant to do. His upstream comments were based only on the mount table info. case and it was an upstream discussion, I don't remember us discussing that internally with Miklos (also see below the comment about the design being different to the usual approach).

I wasn't trying to imply that there was anything non-official going on. The "internal" was really just referring to the fact that in my mind mainly Red Hat fs people were involved in the upstream discussion. The main portion generally agreeing on the direction and keep pointing out that this has been discussed for a long time on-list and during LFSMM etc. and then someone else from the same company strongly disagreeing with the overall design; not just an aspect of it. This is just what it looked like to me taking part in the discussion. Over 2 or almost 3 merge windows David kept sending out the patchset for review. During the review cycle some did review it but every time when the pull request came it got blocked by - sometimes rather sudden - disagreements on the overall design.

**Comment by raven-au** [comment]
> 
> I wasn't trying to imply that there was anything non-official going on. The "internal" was really just referring to the fact that in my mind mainly Red Hat fs people were involved in the upstream discussion. The main portion generally agreeing on the direction and keep pointing out that this has been discussed for a long time on-list and during LFSMM etc. and then someone else from the same company strongly disagreeing with the overall design; not just an aspect of it. This is just what it looked like to me taking part in the discussion. Over 2 or almost 3 merge windows David kept sending out the patchset for review. During the review cycle some did review it but every time when the pull request came it got blocked by - sometimes rather sudden - disagreements on the overall design.

Right, understood, and although it must have sounded like that I didn't actually mean to imply it.

It was a bit surprising Miklos waited so long to comment.

Perhaps it was the fact that the work was essentially done as an upstream project and there happened to be a few people in RedHat that were interested in it, such as myself, who contributed by using it and testing it and sent occasional fix patches to David.

So I think Miklos missed out on contributing to the design early on and that was probably a mistake made by us (being the people we knew about that cared about it).

Nevertheless I'm not sure it would have ended differently had we approached it differently ...

Perhaps if Mikos had posted a cut down proof-of-concept so I could have seen what his idea looked like I may have been more receptive to it or at least it would have enabled me to point out why I thought Davids approach was better (since that's what I thought given his description of what he wanted to do). But there was no secret that I don't particularly like the proc file system because every time I've needed to work on problems relating to it I find it (IMO) convoluted and difficult to follow. The fact is that when testing Davids code working out what was going on was pretty straight forward which I found refreshing.

**Comment by bluca** [comment]
Thanks for the details - so, is there any plan to try and solve these issues and propose the series (or a variation of it) again?

**Comment by raven-au** [comment]
> Thanks for the details - so, is there any plan to try and solve these issues and propose the series (or a variation of it) again?

Not sure, I had asked David some time ago but it wasn't clear what he planned on doing, I'll need to catch up with him again but changes along the lines of what Miklos proposed will make it very difficult to achieve what's needed.

It would be a good idea for Christian to catch with him too I think, the more thoughts and ideas the better I think.

**Comment by karelzak** [comment]
@raven-au from my point of view, David's interface was too complex and with predefined structs. I can imagine a simple fsinfo() interface where userspace defines an array of variables, allocate space for the result and kernel fill the fields by data from VFS; just simple 'key' and 'value' API where one syscall can return multiple values for the mountpoint. This solution will be open for arbitrary new fields and userspace will get who it asks for. I agree that procfs based solution (Mikos' idea?) is odd. It seems better to get so basic information without /proc.

**Comment by poettering** [comment]
So I think what we should be doing to fix the issue is to make unit_start() for mount units return -EAGAIN when we notice the mount table watching is ratelimited. This should fix the job queue issues, as it means we won't act on mount units unless we also can read the mount tab.

It's probably a very simple fix: implement the existing `test_start_limit()` vtable function for .mount units (it currently is only implemented in service units). There, just return EAGAIN when the rate limit of the /proc/self/mountinfo event source is hit.

(The vtable function should probably be renamed though, since it is used for more that start limits.)

In addition to this, once we leave the rate limiting we should issue  `sd_event_source_set_enabled(j->manager->run_queue_event_source, SD_EVENT_ONESHOT)` once. This ensure the run qeue is triggered again, and we try to find runnable jobs again, which might then succeed given the EAGAIN situation was now corrected.

**Comment by Geass-LL** [comment]
Hi, @Vogtinator . Did you check if systemd-250 fix this issure? I have test systemd-250 and find /mnt is still not mounted.

After ratelimit expired, mnt.mount is still active, and systemd will not start an active unit, the returned error is `-EALREADY`. The PR only changes mnt.mount to inactive

**Comment by Vogtinator** [comment]
I haven't tried systemd 250 yet. I'll do that once it arrives in Tumbleweed or maybe earlier if I can find some time to build it myself.

**Comment by Geass-LL** [comment]
> Finally got around to test this. Looks like systemd 250.4 does not fix the issue here, the reproducer still has the same broken result.

Thanks for the test.

**Comment by Vogtinator** [comment]
Still not reopened, ping.

**Comment by msekletar** [comment]
Let's reopen it and I will have a look what is going on.

**Comment by karelzak** [comment]
fsinfo() is dead thing, but since 6.8 we have statmount() and listmount() https://lwn.net/Articles/950569/ 
https://github.com/torvalds/linux/blob/master/include/uapi/linux/mount.h#L143
